### PR TITLE
write to local lindi json file (3)

### PR DESCRIPTION
**Pull request by magland**

### CODE REVIEW

**Comment by magland** [comment]
The goal here is to rearrange things a bit so that we can open .lindi.json file in write modes ("r+", "w", "w-", "x", "a") in the same way that we would open a .h5 file for writing. We don't want to be completely rewriting the .json file with each change... so instead we only rewrite it when `flush()` is called (which is an h5py function)... and note that `flush()` is also called when the file is closed (including when it goes out of context).

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 1]
```diff
@@ -190,6 +232,53 @@ def to_reference_file_system(self):
...
+        rfs = self.to_reference_file_system()
+        blobs_to_upload = set()
+        for k, v in rfs['refs'].items():
+            if isinstance(v, list) and len(v) == 3:
+                url = _apply_templates(v[0], rfs.get('templates', {}))
+                if not url.startswith("http://") and not url.startswith("https://"):
+                    local_path = url
+                    blobs_to_upload.add(local_path)
+        blob_mapping = _upload_blobs(blobs_to_upload, on_upload_blob=on_upload_blob)
+        for k, v in rfs['refs'].items():
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:276]:
```suggestion
        # Replace the local URLs in rfs['refs'] with URLs of the uploaded files
        for k, v in rfs['refs'].items():
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 2]
```diff
@@ -190,6 +232,53 @@ def to_reference_file_system(self):
...
+
+        Returns
+        -------
+        str
+            The URL (or local path) of the uploaded reference file system .json
+            file.
+        """
+        rfs = self.to_reference_file_system()
+        blobs_to_upload = set()
+        for k, v in rfs['refs'].items():
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:266]:
```suggestion
        # Get the set of all local URLs in rfs['refs']
        for k, v in rfs['refs'].items():
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 3]
```diff
@@ -190,6 +232,53 @@ def to_reference_file_system(self):
...
+            if isinstance(v, list) and len(v) == 3:
+                url = _apply_templates(v[0], rfs.get('templates', {}))
+                if not url.startswith("http://") and not url.startswith("https://"):
+                    local_path = url
+                    blobs_to_upload.add(local_path)
+        blob_mapping = _upload_blobs(blobs_to_upload, on_upload_blob=on_upload_blob)
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:274]:
```suggestion
        # Upload each of the local blobs using the given upload function and get a mapping from
        # the original file paths to the URLs of the uploaded files
        blob_mapping = _upload_blobs(blobs_to_upload, on_upload_blob=on_upload_blob)
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 4]
```diff
@@ -190,6 +232,53 @@ def to_reference_file_system(self):
...
+        for k, v in rfs['refs'].items():
+            if isinstance(v, list) and len(v) == 3:
+                url1 = _apply_templates(v[0], rfs.get('templates', {}))
+                url2 = blob_mapping.get(url1, None)
+                if url2 is not None:
+                    v[0] = url2
+        with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:283]:
```suggestion
        # Write the updated LINDI file to a temp directory and upload it
        with tempfile.TemporaryDirectory() as tmpdir:
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 5]
```diff
@@ -218,7 +307,7 @@ def driver(self):
 
     @property
     def mode(self):
-        return self._mode
+        return 'r' if self._mode == 'r' else 'r+'
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:317]: Why is this needed?

Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:317]: You would think it should return self._mode. But actually, h5py only returns r or r+. See

https://github.com/h5py/h5py/blob/b2a83a03d9850076a427685a488ceeb0f77c5bd0/h5py/_hl/files.py#L318-L323

So we need to do this to be consistent.

Comment (rly) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:317]: That's odd. OK

**Review by rly — APPROVED** [review]
